core/store/src/trie/mem/flexible_data/mod.rs

```diff
@@ -0,0 +1,65 @@
+use super::arena::{ArenaSlice, ArenaSliceMut};
+
+pub mod children;
+pub mod encoding;
+pub mod extension;
+pub mod value;
+
+/// This trait simplifies our programming against flexibly-sized structures.
+/// A flexibly-sized structure is one whose memory layout depends on runtime
+/// values. For example, we may use a flexibly-sized structure to encode two
+/// variable-sized strings, like this:
+///
+///   0                4                 8         8+str1len    8+str1len+str2len
+///   |  str1 length   |   str2 length   |    str1    |     str2    |
+///
+/// Why do we do this? Because this is a lot more memory-efficient than the
+/// alternative of storing two heap-allocated Strings; it is a single memory
+/// allocation rather than three, and it saves the need to encode two pointers.
+///
+/// There's no Rust-safe way to do this, however, and it would be too error-
+/// prone to encode and decode these structures manually. This is where
+/// FlexibleDataHeader comes in.
+///
+/// In this string example, these two strings would each define a header struct,
+/// say EncodedStringHeader, that implements FlexibleDataHeader. The struct
+/// itself only contains the fixed-size part, i.e. the length of the string.
+/// Then, we implement encode_flexible_data and decode_flexible_data to specify
+/// how the flexibly-sized part of the data is encoded and decoded. These, of
+/// course need to be consistent with each other.
+///
+/// This trait allows us to then encode and decode a flexibly-sized structure
+/// with multiple flexibly-sized parts with relative ease.
+///
+/// It is possible to envision a more general way to do this that is also safer,
+/// but for now, we'll keep it simple.
```

Comment:
I think this comment is no longer relevant, now that FlexibleDataHeader is basically safe :)